apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/CoreConfiguration.java

```diff
@@ -777,25 +777,26 @@ public TimeDuration getSpanMinDuration() {
      * Makes sure to not initialize ConfigurationOption, which would initialize the logger
      */
     @Nullable
-    public static String getConfigFileLocation(List<ConfigurationSource> configurationSources) {
-        String configFileLocation = DEFAULT_CONFIG_FILE;
-        for (ConfigurationSource configurationSource : configurationSources) {
-            String valueFromSource = configurationSource.getValue(CONFIG_FILE);
-            if (valueFromSource != null) {
-                configFileLocation = valueFromSource;
+    public static String getConfigFileLocation(List<ConfigurationSource> configurationSources, boolean premain) {
+        String configFileLocation = premain ? DEFAULT_CONFIG_FILE : null;
+        for (ConfigurationSource source : configurationSources) {
+            String value = source.getValue(CONFIG_FILE);
+            if (value != null) {
+                configFileLocation = value;
                 break;
             }
         }
-        if (configFileLocation.contains(AGENT_HOME_PLACEHOLDER)) {
-            String agentHome = ElasticApmAgent.getAgentHome();
-            if (agentHome != null) {
-                return configFileLocation.replace(AGENT_HOME_PLACEHOLDER, agentHome);
-            } else {
-                return null;
-            }
-        } else {
+
+        if (configFileLocation == null || !premain || !configFileLocation.contains(AGENT_HOME_PLACEHOLDER)) {
             return configFileLocation;
         }
```

Comment:
It's confusing and I don't think necessary.
At this point, if `configFileLocation != null` always return `configFileLocation.replace(AGENT_HOME_PLACEHOLDER, agentHome)`, otherwise return `null`

Reply:
Should be simplified now.